### DISCUSSION TITLE
Is it possible to make menu with OLED screen + buttons? Like Gnarbox2

### DISCUSSION THREAD

**Discussion by Stefankante** [outdoorbits/little-backup-box#202] [Ideas]
Sometime accessing on phone is not an option (most of the time).
I wondered if It could make menus system for OLED screen and make it execute such command like on web UI.
A simple 4 way buttons are quite nice.

Outdoorbits: Sorry, I just removed the gnarbox images, most probably they are not free at all. ;-)

**Comment by outdoorbits** [comment]
Of course, this might technically not be a serious problem. I really 
would like to code this. But there is another problem before: We need a 
case. I don't know any case including a display and buttons.

Of course, there are stl-files in this repository. But unfortunately 
stl-files are static, not made for modifications. A very important first 
step would be to create a case constructed in an open source environment 
like https://www.freecad.org . From this we could fork easily variants for different 
displays and such ones for buttons.

I tried to do this but failed in lack of time and construction skills. 
There is a nice tutorial, even containing measures, but unfortunately in 
German only: https://www.youtube.com/watch?v=mTOtj-k5WJk . Are you maybe 
able to draw this? Or anybody else?

**Reply by Stefankante** [reply]
There is folk of LBB from 2 yrs ago has interfacing with buttons, [https://github.com/gregmag/little-backup-box ](url)
I know it has 2 buttons but you could actually make web ui config like OLED variants by making users to choose GPIOs according to their buttons 

- UP position = GPIO XX as drop down menu (those which could by supported as input.
which will remove all hassle to find suitable board for it. 

About case, I'm not technically prep for case currently as the enclosure will follow the function. As you saw my Pi4 has heatsink wrap around it and I just stick OLED on the heatsink lol. 

The buttons I knew the pain to get decent one and looking good, But I think this one work best to get same case with just minor drilling.
This membrane switches are cheap and already labeled on buttons which I think layout make sense for our use.
[https://th.aliexpress.com/item/32917032659.html?spm=a2g0o.productlist.main.19.1478246cGbygeN&algo_pvid=31600896-621d-4afb-be1b-89561ccc6a33&aem_p4p_detail=202304070555032198781137240370001082396&algo_exp_id=31600896-621d-4afb-be1b-89561ccc6a33-9&pdp_npi=3%40dis%21THB%2191.66%2191.66%21%21%21%21%21%40211be72e16808721035714235d07b6%2166059244162%21sea%21TH%21143148292&curPageLogUid=7ovvqQFayNxf&ad_pvid=202304070555032198781137240370001082396_2&ad_pvid=202304070555032198781137240370001082396_2](url)

![61zkryDBbeL](https://user-images.githubusercontent.com/35141284/230611223-95a84703-ce65-4b97-8bc9-fd566fe0ce97.jpg)

I have use similar thing with clear pi case (ESP32 inside) with OLED stick inside case and drill just tiny slot for that membrane switch cable to get inside the case.
![tempImagewlh1lK](https://user-images.githubusercontent.com/35141284/230611804-fe98ef07-57c6-4d7e-95e4-a69a91504c6b.gif)

**Reply by Stefankante** [reply]
I'll get that STL printed to get exactly same setup as others.

**Comment by outdoorbits** [comment]
Creating a menu will not be a quick task but will take many hours. So 
for me, I have to say, the function follows the case. ;-) I don't want 
to spend that many time not knowing if there ever will be a suitable 
hardware to use it.  It's very important to me, having a traveling proof 
solution, meaning a protective case. But maybe somebody else will create 
the case?

**Reply by Stefankante** [reply]
Photoshopped, BTW the button is quite hard to press accidentally.
![91dc2af9-6452-4db6-8fee-47006d11ec9b copy](https://user-images.githubusercontent.com/35141284/230616150-d960bb4e-3e18-4edf-a34c-02a1b3ff76a1.jpg)

**Reply by Stefankante** [reply]
I wondered if I have 3D design skill.

**Comment by outdoorbits** [comment]
I couldn't stop thinking about, so every minute I found, I worked on the menu. ;-)
There are still a few hours to code before publishing a first version...
It will be driven by four buttons, probably GPIO 31, 33, 35 and 37. Action on rising edge...

**Comment by outdoorbits** [comment]
I just uploaded a new version including this functionality. You will need 4 Buttons, more in the [Wiki](https://github.com/outdoorbits/little-backup-box/wiki/02a.-Displaymenu).

**Comment by Fabesson** [comment]
Is there a reason you didn't use the GPIO 31, 33, 35 and 37 as originally planned?
Because with the membrane buttons it could be possible to connect the 5 pin connector directly to the gpios
to the 5 pins (31, 33, 35 and 37) and GND (39). The detection then would be the falling edge.
Unfortunately I don't have such membrane buttons at home for testing.

**Reply by outdoorbits** [reply]
I'm close to upload configurable buttons. Do you know the witch button function should be connected to each pin?

**Reply by Fabesson** [reply]
I don‘t know which button is connected to which pin in the 5-pin connector. I still don‘t have the membrane buttons at home for testing.

**Comment by outdoorbits** [comment]
This was just because there are other functions mentioned for PIN  33 
and 35. Of course, they are not in use by lbb...

I did another try to construct a case. This should work using this 
buttons: 
https://www.amazon.de/Einbau-Taster-Drucktaster-Schalter-Klingel/dp/B08F48GQPW/ref=asc_df_B08F48GQPW/?tag=googshopde-21&linkCode=df0&hvadid=579313501848&hvpos=&hvnetw=g&hvrand=16846023521108110301&hvpone=&hvptwo=&hvqmt=&hvdev=c&hvdvcmdl=&hvlocint=&hvlocphy=9041978&hvtargid=pla-1653354096431&th=1

I hope to finish it within a few days.

**Reply by Fabesson** [reply]
Okay, thank you for the furter information.
I have buttons like this at home from several projects, but I think these will not fit into my case.
I use a RPi 3B+ and a transparent case from RS Components and a glued display on the inside.
But maybe if the RPi 4 is better available, I will switch to the PRi 4 and the Case you created.
For my needs i added two smd buttons at the gpios with a python script for shutdown command and the "comitup-cli d" command.
Something link this:
[https://www.reichelt.de/kurzhubtaster-6x6mm-hoehe-4-3mm-12v-vertikal-taster-3301-p27892.html?&trstct=pos_14&nbc=1](https://www.reichelt.de/kurzhubtaster-6x6mm-hoehe-4-3mm-12v-vertikal-taster-3301-p27892.html?&trstct=pos_14&nbc=1)

**Comment by outdoorbits** [comment]
If this would solve a problem, I could make this configurable, so you could select your combination of GPIOs.

**Reply by Fabesson** [reply]
I know this is an extra todo, but I think for the most users it would be easier if the GPIOs are individually adjustable.
@Stefankante also mentioned this some days ago.
Would you only let the users choose between free gpio ports or would you add all gpios?
I actually don't know which GPIOs are already used. I only got some errors/notifications on some GPIOs in python when trying to add an "add_event_detect".

**Comment by outdoorbits** [comment]
My idea was to offer sets of GPIOs, else I had to check for double or 
disfunctional combinations.

**Reply by Fabesson** [reply]
Sounds like a good idea and much easier to implement.

**Comment by vitotai** [comment]
Why not use this HAT. It is a SH1106 based OLED, so it should work.
https://www.aliexpress.com/item/1005002741139840.html

I would prefer this one with TFT LCD. My experience with OLED isn't very good. They are prone to failing.
https://www.aliexpress.com/item/32970035492.html
The controller of the LCD is supported by luma.lcd, so it might be possible, isn't it?

**Comment by outdoorbits** [comment]
Looks good. If you need, I can add the pin-configuration-option. Just 
tell me then.

**Comment by ruebyi** [comment]
Ou guys are awesome!

**Comment by outdoorbits** [comment]
There is another project on its way: I created a new case for a colour display, buttons and a cooler. I'm still in reworking...
